crates/slightfile/src/capability_store.rs

```diff
@@ -0,0 +1,177 @@
+use std::collections::HashMap;
+
+use crate::ResourceName;
+
+type CapabilityName = String;
+
+/// A store for slight capabilities.
+///
+/// The inner structure of this store is a nested HashMap.
+/// The outer HashMap is keyed by the name of the capability. e.g. "keyvalue" or "messaging"
+/// The inner HashMap is keyed by the name of the resource. e.g. "keyvalue.redis" or "messaging.mosquitto"
+///
+/// The `get` function returns the first resource with the given name living in its namespace.
+/// If a resource uses the Any resource name, any name can be used to retrieve it.
+///
+/// Usage:
+///
+/// ```rust
+/// use slight_file::slightfile::CapabilityStore;
+/// use slight_file::slightfile::ResourceName;
+///
+/// pub struct State;
+///
+/// let mut store = CapabilityStore::new();
+/// let mut state = State;
+/// store.insert(ResourceName::Specific("keyvalue.redis"), "redis", state);
+/// store.insert(ResourceName::Any, "messaging", state);
+///
+/// assert_eq!(store.get("keyvalue.redis"), state);
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct CapabilityStore<T> {
+    inner: HashMap<CapabilityName, HashMap<ResourceName, T>>,
```

Comment:
I think `ResourceName` and `CapabilityName` can be easily confused. Because the fields in the slight file are `resource`, and `name` respectively. Why don't we just call them `Resource`/`Name`, or, `CapabilityResource`/`CapabilityName`?

Reply:
Umm it's actually not clear to me what names are the best to choose from for thesse variables. Let me try to list different kinds of "names" we have from the slightfile

1. "my-container" is the name for a resource that apps want to refer to when they do `keyvalue::open(.)`
2. "keyvalue" is a name for a capability
3. "keyvalue.redis" is a name for resource

Perhaps we should call 1. as `ResourceName`, 2. as `CapabilityKind` and 3. as `ResourceKind`?

Reply:
How about:
"my-container") "CapabilityGuestName"
"keyvalue") "CapabilityType"
"keyvalue.redis") "CapabilityImplementor"
... ?

Reply:
I like `CapabilityType` more than `CapabilityKind` and `capabilityimplementor` more than `resourcekind`. 

I don't like to mention `Guest` in the slightfile crate as it is designed from the guests' perspectives. I think `resource` is a good word for describing something apps want to get when doing `keyvalue.open` or `messaging::Pub::open()` in a sense that it is acquiring something it owns.